### ISSUE TITLE
[ECP-8486][ECP-8595] Grand total without tax reported to GPay, Apple Pay

### ISSUE THREAD

**Issue by brain-effect** [Adyen/adyen-magento2-express-checkout#37]
**Describe the bug**
Magento 2.4.4-p4 + Adyen Express Checkout 1.1.2 + Adyen Payment 8.19.0. Adyen in Test mode.
The grand total without the tax included is reported to GPay and Apple Pay as the final amount. The final amount reported to Adyen, however, will be the grand total with tax, which will (most likely?) cause errors in live system.

**To Reproduce**
1. Enable express method buttons, for example in minicart.
2. Add any taxable product to cart.
3. Click on GPay or Apple Pay express buttons.
4. The grand total without tax is reported to GPay and Apple Pay as the final amount.

**Expected behavior**
The grand total with tax included should be reported to GPay and Apple Pay as the final amount.

**Screenshots**
![Screenshot 2023-07-03 at 1 42 41 PM](https://github.com/Adyen/adyen-magento2-express-checkout/assets/95756740/29a582bc-9928-42d4-a780-832e9b5257a6)

**Desktop (please complete the following information):**
 - Any OS
 - Any browser

**Smartphone (please complete the following information):**
 - Any Device
 - Any OS
 - Any Browser
 - Any Version

**Additional context**
Probably related to https://github.com/magento/magento2/issues/34649

**Comment by candemiralp** [comment]
Hello @brain-effect,

Thank you for raising this issue. We've created a PR for Apple Pay tax calculation on express payment [here](https://github.com/Adyen/adyen-magento2-express-checkout/pull/38). We are currently working on Google Pay fix.

Best Regards,
Can

**Comment by smiroshnyk** [comment]
Hello.
It still doesn't work even for Apple Pay.

![photo_2023-09-11 15 55 37](https://github.com/Adyen/adyen-magento2-express-checkout/assets/20116324/7c37a32b-330a-4e5c-843b-2069e80da05b)
![image](https://github.com/Adyen/adyen-magento2-express-checkout/assets/20116324/f166d704-3f6e-46fd-914f-9be13cedcb7d)

You changed it in one place, but you set grant_total in 4 places.

The steps to reproduce are the same as in this issue description.

**Comment by joselopr** [comment]
@candemiralp

**Comment by candemiralp** [comment]
Hello @joselopr,

Thank you mentioning the consequent issues related to Apple Pay express and apologies for the late response. We will test Apple Pay tax issue once again. Let me update you soon.

Best Regards,
Can

**Comment by dimitriBouteille** [comment]
Hi @candemiralp @raoulritter  Why in version 1.2.0 the fix is not visible ? 

**In 1.1.3 :**

```js
applePayShippingContactUpdate.newTotal = {
    label: $t('Grand Total'),
    amount: (response.grand_total + response.tax_amount).toString()
};
```

**In 1.2.0 :**

```js
applePayShippingContactUpdate.newTotal = {
    label: this.getMerchantName(),
    amount: (response.grand_total).toString()
};
```

I have this error when I try to place order with version 1.2.0 

```json
{
   "status" : 422,
   "errorCode" : "5_001",
   "message" : "ApplePay token amount-mismatch",
   "errorType" : "validation",
   "pspReference" : "R3TB5Q4HKJHKGK82"
}
```

It's possible to fix this issue quickly ? :heart:

**Comment by bitumin** [comment]
In case it helps anyone, these are the patches I currently use (Magento 2.4.4 + Adyen 9.12.2 + Adyen ExpressCheckout 2.4.1). Be advised that using base grand total instead of grand total might NOT be a good replacement in shops with multiple currencies (feel free to find a better alternative for those).

ApplePay Button
```
diff --git a/vendor/adyen/adyen-magento2-expresscheckout/view/frontend/web/js/applepay/button.js b/vendor/adyen/adyen-magento2-expresscheckout/view/frontend/web/js/applepay/button.js
--- a/vendor/adyen/adyen-magento2-expresscheckout/view/frontend/web/js/applepay/button.js
+++ b/vendor/adyen/adyen-magento2-expresscheckout/view/frontend/web/js/applepay/button.js
@@ -418,7 +418,7 @@
                 applePayShippingMethodUpdate.newTotal = {
                     type: 'final',
                     label: this.getMerchantName(),
-                    amount: (response.grand_total).toString()
+                    amount: (response.base_grand_total).toString()
                 };

                 // If the shipping methods is an array pass all methods to Apple Pay to show in payment window.

```

GooglePay Button
```
diff --git a/vendor/adyen/adyen-magento2-expresscheckout/view/frontend/web/js/googlepay/button.js b/vendor/adyen/adyen-magento2-expresscheckout/view/frontend/web/js/googlepay/button.js
--- a/vendor/adyen/adyen-magento2-expresscheckout/view/frontend/web/js/googlepay/button.js
+++ b/vendor/adyen/adyen-magento2-expresscheckout/view/frontend/web/js/googlepay/button.js
@@ -405,7 +405,7 @@
                                         ],
                                         currencyCode: totals.quote_currency_code,
                                         totalPriceStatus: 'FINAL',
-                                        totalPrice: (totals.grand_total).toString(),
+                                        totalPrice: (totals.base_grand_total).toString(),
                                         totalPriceLabel: 'Total',
                                         countryCode: configModel().getConfig().countryCode
                                     }

```

**Comment by bitumin** [comment]
Magento has a fix for this for >= 2.4.6 apparently, the relevant commits are these:

https://github.com/magento/magento2/commit/ff39d2a0b4ae2a6412e3f123c9d0cef77c6d5fca#diff-bccd9ecb980856e5bfeaf0c83df24b0852b06e4900c9e8532b91bbade385c3d2

https://github.com/magento/magento2/commit/67f6f468e319c3b442bd49d5c5fe5e704c21f251

It should be easy to create a patch with those commit changes, and in that case there would be no need to patch adyen module (I have not tested it yet, though).

**Comment by candemiralp** [comment]
Hello all,

I am closing this thread as this has been fixed earlier on #45 and #38.

Best Regards,
Can
